src/Language/Marlowe/Core/V1/Semantics.purs

```diff
@@ -183,6 +189,44 @@ evalObservation env state obs =
       TrueObs -> true
       FalseObs -> false
 
+-- | Reduce a Contract
+reduceContract :: Contract -> Contract
+reduceContract contract =
+  let (ident :: Identity Contract) = reduceContract' contract in unwrap ident
+
+reduceContract' :: forall m. Monad m => Contract -> m Contract
+reduceContract' = rewriteContractTopDown visitor >=> rewriteContractBottomUp
+  visitor
+  where
+  visitor = Visitor { onCase, onContract, onObservation, onValue }
+
+  onCase = pure
+  onContract = pure
+  onObservation = pure
+
+  onValue (NegValue (NegValue a)) = onValue a
+  onValue (NegValue (SubValue a b)) = onValue $ SubValue b a
+  onValue (AddValue a b) | b == zero = onValue a
+  onValue (AddValue a b) | a == zero = onValue b
+  onValue (SubValue a b) | b == zero = onValue a
+  onValue (SubValue a b) | a == zero = onValue $ NegValue b
+  onValue (MulValue a b) | b == one = onValue a
+  onValue (MulValue a b) | a == one = onValue b
+  onValue (MulValue _ b) | b == zero = pure zero
+  onValue (MulValue a _) | a == zero = pure zero
+  onValue (DivValue a b) | b == one = onValue a
+  onValue (DivValue a _) | a == zero = pure zero
+  onValue (MulValue a (DivValue b c)) | a == c = onValue b
+  onValue (MulValue (DivValue b c) a) | a == c = onValue b
+  onValue (MulValue (DivValue b c) a) | a == b = onValue c
+  onValue (DivValue (MulValue b c) a) | a == c = onValue b
+  onValue (DivValue (MulValue b c) a) | a == b = onValue c
+
+  onValue v = pure v
```

Comment:
could we rename and move these? 
Instead of `reduceContract` I would call it `simplifyContract`, we already use the word reduce in a different way in this file.

Also instead of `Semantics.purs` could this be in a `Compiler`/`Performance`/`Simplification`/`Utilities` package? I'd like to avoid increasing this package size if possible (and at some point for a future release I'd like a global refactor in all modules that changes Semantics => Interpreter)

Reply:
Also, it probably makes sense to have the onValue simplification as a top level function, that way you can make tests (directly or via the test-spec tool) easier.

Reply:
In the specification of the Marlowe semantics there are similar reduction/simplification rules: https://github.com/input-output-hk/marlowe/blob/6afc2621128625088aff7664bdb6ded5897c5e67/isabelle/Core/Semantics.thy#L78-L218
We should probably port the missing ones as well.